diff -Naur a/drivers/net/ethernet/broadcom/tg3.c b/drivers/net/ethernet/broadcom/tg3.c
--- a/drivers/net/ethernet/broadcom/tg3.c	2019-03-27 13:13:56.000000000 +0800
+++ b/drivers/net/ethernet/broadcom/tg3.c	2019-04-26 11:08:20.307848693 +0800
@@ -11734,6 +11734,12 @@
 		pci_set_power_state(tp->pdev, PCI_D3hot);
 	}

+	if (tg3_asic_rev(tp) == ASIC_REV_5720){
+		/*Fixed ASIC_REV_5720 support 100M/10M feature */
+       __tg3_writephy(tp, 0x8, 0x10, 0x1d0);
+       __tg3_writephy(tp, 0x1f, 0x4, 0x5e1);
+	}
+
 	return err;
 }

