--- /dev/null
+++ b/arch/arm/mach-iproc/include/mach/nand_iproc.h
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2013, Broadcom Corporation. All Rights Reserved.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef _NAND_IPROC_H_
+#define _NAND_IPROC_H_
+
+/*
+ * Registers used directly in driver
+ */
+
+/*
+ * Shared Structure
+ */
+struct mtd_partition;
+struct brcmnand_platform_data {
+	int			chip_select;
+	int			strap_boot;
+	int			strap_type;
+	int			strap_page_size;
+	int			nr_parts;
+	struct mtd_partition	*parts;
+};
+
+#endif /* _NAND_IPROC_H_ */
diff --git a/arch/arm/mach-iproc/include/mach/qspi_iproc.h b/arch/arm/mach-iproc/include/mach/qspi_iproc.h
new file mode 100644
index 0000000..14fcac8
