--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -545,6 +545,19 @@ config I6300ESB_WDT
 	  To compile this driver as a module, choose M here: the
 	  module will be called i6300esb.
 
+config IE6XX_WDT
+        tristate "Intel Atom E6xx Watchdog"
+        depends on X86 && PCI
+        select WATCHDOG_CORE
+        select MFD_CORE
+        select LPC_SCH
+        ---help---
+          Hardware driver for the watchdog timer built into the Intel
+          Atom E6XX (TunnelCreek) processor.
+
+          To compile this driver as a module, choose M here: the
+          module will be called ie6xx_wdt.
+
 config INTEL_SCU_WATCHDOG
 	bool "Intel SCU Watchdog for Mobile Platforms"
 	depends on X86_MRST
diff --git a/drivers/watchdog/Makefile b/drivers/watchdog/Makefile
index fe893e9..8f591c5 100644
