--- a/include/linux/i2c/at24.h
+++ b/include/linux/i2c/at24.h
@@ -3,6 +3,7 @@
 
 #include <linux/types.h>
 #include <linux/memory.h>
+#include <linux/eeprom_class.h>
 
 /*
  * As seen through Linux I2C, differences between the most common types of I2C
@@ -24,9 +25,11 @@ struct at24_platform_data {
 #define AT24_FLAG_READONLY	0x40	/* sysfs-entry will be read-only */
 #define AT24_FLAG_IRUGO		0x20	/* sysfs-entry will be world-readable */
 #define AT24_FLAG_TAKE8ADDR	0x10	/* take always 8 addresses (24c00) */
+#define AT24_FLAG_DISABLE_I2CBLOCK 0x08 /*disable smbus i2c block access */
 
 	void		(*setup)(struct memory_accessor *, void *context);
 	void		*context;
+	struct eeprom_platform_data *eeprom_data; /* extra data for the eeprom_class */
 };
 
 #endif /* _LINUX_AT24_H */
diff --git a/include/linux/i2c/pca953x.h b/include/linux/i2c/pca953x.h
index 139ba52..2a69244 100644
