--- a/include/linux/ceph/messenger.h
+++ b/include/linux/ceph/messenger.h
@@ -92,7 +92,11 @@ struct ceph_msg {
 	bool front_is_vmalloc;
 	bool more_to_follow;
 	bool needs_out_seq;
+#ifdef __GENKSYMS__
+	int front_max;			/* renamed */
+#else
 	int front_alloc_len;
+#endif
 	unsigned long ack_stamp;        /* tx: when we were acked */
 
 	struct ceph_msgpool *pool;
diff --git a/include/linux/connector.h b/include/linux/connector.h
index 3c9c54f..7638407 100644
