--- a/include/linux/percpu.h
+++ b/include/linux/percpu.h
@@ -73,9 +73,9 @@
  * intelligent way to determine this would be nice.
  */
 #if BITS_PER_LONG > 32
-#define PERCPU_DYNAMIC_RESERVE		(20 << 10)
+#define PERCPU_DYNAMIC_RESERVE		(28 << 10)
 #else
-#define PERCPU_DYNAMIC_RESERVE		(12 << 10)
+#define PERCPU_DYNAMIC_RESERVE		(20 << 14)
 #endif
 
 extern void *pcpu_base_addr;
diff --git a/include/linux/perf_event.h b/include/linux/perf_event.h
index 8d5b91e..aaed55b 100644
