--- a/include/linux/mtd/cfi_endian.h
+++ b/include/linux/mtd/cfi_endian.h
@@ -37,35 +37,48 @@
 #define CFI_BIG_ENDIAN
 #endif
 
+#ifdef CONFIG_MTD_CFI_OF_BYTE_SWAP
+#define CFI_OF_BYTE_SWAP
+#endif
+
 #endif
 
 #if defined(CFI_LITTLE_ENDIAN)
-#define cpu_to_cfi8(x) (x)
-#define cfi8_to_cpu(x) (x)
-#define cpu_to_cfi16(x) cpu_to_le16(x)
-#define cpu_to_cfi32(x) cpu_to_le32(x)
-#define cpu_to_cfi64(x) cpu_to_le64(x)
-#define cfi16_to_cpu(x) le16_to_cpu(x)
-#define cfi32_to_cpu(x) le32_to_cpu(x)
-#define cfi64_to_cpu(x) le64_to_cpu(x)
+#define cpu_to_cfi8(m,x) (x)
+#define cfi8_to_cpu(m,x) (x)
+#define cpu_to_cfi16(m,x) cpu_to_le16(x)
+#define cpu_to_cfi32(m,x) cpu_to_le32(x)
+#define cpu_to_cfi64(m,x) cpu_to_le64(x)
+#define cfi16_to_cpu(m,x) le16_to_cpu(x)
+#define cfi32_to_cpu(m,x) le32_to_cpu(x)
+#define cfi64_to_cpu(m,x) le64_to_cpu(x)
 #elif defined (CFI_BIG_ENDIAN)
-#define cpu_to_cfi8(x) (x)
-#define cfi8_to_cpu(x) (x)
-#define cpu_to_cfi16(x) cpu_to_be16(x)
-#define cpu_to_cfi32(x) cpu_to_be32(x)
-#define cpu_to_cfi64(x) cpu_to_be64(x)
-#define cfi16_to_cpu(x) be16_to_cpu(x)
-#define cfi32_to_cpu(x) be32_to_cpu(x)
-#define cfi64_to_cpu(x) be64_to_cpu(x)
+#define cpu_to_cfi8(m,x) (x)
+#define cfi8_to_cpu(m,x) (x)
+#define cpu_to_cfi16(m,x) cpu_to_be16(x)
+#define cpu_to_cfi32(m,x) cpu_to_be32(x)
+#define cpu_to_cfi64(m,x) cpu_to_be64(x)
+#define cfi16_to_cpu(m,x) be16_to_cpu(x)
+#define cfi32_to_cpu(m,x) be32_to_cpu(x)
+#define cfi64_to_cpu(m,x) be64_to_cpu(x)
 #elif defined (CFI_HOST_ENDIAN)
-#define cpu_to_cfi8(x) (x)
-#define cfi8_to_cpu(x) (x)
-#define cpu_to_cfi16(x) (x)
-#define cpu_to_cfi32(x) (x)
-#define cpu_to_cfi64(x) (x)
-#define cfi16_to_cpu(x) (x)
-#define cfi32_to_cpu(x) (x)
-#define cfi64_to_cpu(x) (x)
+#define cpu_to_cfi8(m,x) (x)
+#define cfi8_to_cpu(m,x) (x)
+#define cpu_to_cfi16(m,x) (x)
+#define cpu_to_cfi32(m,x) (x)
+#define cpu_to_cfi64(m,x) (x)
+#define cfi16_to_cpu(m,x) (x)
+#define cfi32_to_cpu(m,x) (x)
+#define cfi64_to_cpu(m,x) (x)
+#elif defined (CFI_OF_BYTE_SWAP)
+#define cpu_to_cfi8(m,x) (x)
+#define cfi8_to_cpu(m,x) (x)
+#define cpu_to_cfi16(m,x) ((m)->byteswap ? swab16(x) : (x))
+#define cpu_to_cfi32(m,x) ((m)->byteswap ? swab32(x) : (x))
+#define cpu_to_cfi64(m,x) ((m)->byteswap ? swab64(x) : (x))
+#define cfi16_to_cpu(m,x) ((m)->byteswap ? swab16(x) : (x))
+#define cfi32_to_cpu(m,x) ((m)->byteswap ? swab32(x) : (x))
+#define cfi64_to_cpu(m,x) ((m)->byteswap ? swab64(x) : (x))
 #else
 #error No CFI endianness defined
 #endif
diff --git a/include/linux/mtd/map.h b/include/linux/mtd/map.h
index 3887901..b8ca7de 100644
