diff -Nu a/drivers/i2c/busses/i2c-ismt.c b/drivers/i2c/busses/i2c-ismt.c
--- a/drivers/i2c/busses/i2c-ismt.c	2017-05-09 23:52:56.680565000 -0700
+++ b/drivers/i2c/busses/i2c-ismt.c	2017-05-10 02:09:53.237489185 -0700
@@ -614,6 +614,12 @@
 	priv->head++;
 	priv->head %= ISMT_DESC_ENTRIES;
 
+	if( ret != 0 )
+	{
+		dev_dbg(dev, "Retry i2c-ismt access\n");
+		return -EAGAIN;
+	}
+
 	return ret;
 }
 
