Index: linux-3.16.39/drivers/mfd/lpc_ich.c
===================================================================
--- linux-3.16.39.orig/drivers/mfd/lpc_ich.c
+++ linux-3.16.39/drivers/mfd/lpc_ich.c
@@ -493,6 +493,7 @@ static struct lpc_ich_info lpc_chipset_i
 	[LPC_LPT] = {
 		.name = "Lynx Point",
 		.iTCO_version = 2,
+		.gpio_version = ICH_V5_GPIO,
 	},
 	[LPC_LPT_LP] = {
 		.name = "Lynx Point_LP",
