diff -rupN a/arch/powerpc/boot/dts/powerpc-dni-7448-r0.dts b/arch/powerpc/boot/dts/powerpc-dni-7448-r0.dts
--- a/arch/powerpc/boot/dts/powerpc-dni-7448-r0.dts	1969-12-31 16:00:00.000000000 -0800
+++ b/arch/powerpc/boot/dts/powerpc-dni-7448-r0.dts	2015-06-04 10:40:17.736835830 -0700
@@ -0,0 +1,488 @@
+/*
+ * D
+ *
+ * Copyright 2013 Big Switch Networks, Inc.
+ *
+ * This program is free software; you can redistribute  it and/or modify it
+ * under  the terms of  the GNU General  Public License as published by the
+ * Free Software Foundation;  either version 2 of the  License, or (at your
+ * option) any later version.
+ */
+
+/dts-v1/;
+
+/ {
+	model = "powerpc-dni-7448-r0";
+	compatible = "delta-7448";
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	aliases {
+		ethernet0 = &enet2;
+		serial0 = &serial0;
+		serial1 = &serial1;
+		pci2 = &pci2;
+		mpic-msgr-block0 = &mpic_msgr_block0;
+		mpic-msgr-block1 = &mpic_msgr_block1;
+	};
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		PowerPC,P2020@0 {
+			device_type = "cpu";
+			reg = <0x0>;
+			next-level-cache = <&l2>;
+		};
+
+		PowerPC,P2020@1 {
+			device_type = "cpu";
+			reg = <0x1>;
+			next-level-cache = <&l2>;
+		};
+	};
+
+	memory {
+		device_type = "memory";
+	};
+
+	localbus@ffe05000 {
+		#address-cells = <2>;
+		#size-cells = <1>;
+		compatible = "fsl,p2020-elbc", "fsl,elbc", "simple-bus";
+		reg = <0 0xffe05000 0 0x1000>;
+		interrupts = <19 2>;
+		interrupt-parent = <&mpic>;
+
+		ranges = <0x0 0x0 0x0 0xe8000000 0x8000000
+			  0x1 0x0 0x0 0xffdf0000 0x0000100>;
+
+                flash@0,0 {
+
+                       #address-cells = <1>;
+                       #size-cells = <1>;
+                       compatible = "cfi-flash";
+                       reg = <0x0 0x0 0x8000000>;
+                       bank-width = <2>;
+                       endian = "little";
+
+                       flash@0 {
+                               reg = <0x00000000 0x07b60000>;
+                               label = "open";
+                       };
+                       flash@1 {
+                               /* 4MB onie */
+                               reg = <0x07b60000 0x00400000>;
+                               label = "onie";
+                       };
+                       flash@2 {
+                               /* 128KB, 1 sector */
+                               reg = <0x07f60000 0x00020000>;
+                               label = "uboot-env";
+                               env_size = <0x2000>;
+                       };
+                       flash@3 {
+                               /* 512KB u-boot */
+                               reg = <0x07f80000 0x00080000>;
+                               label = "uboot";
+                       };
+               };
+
+		cpld@1,0 {
+			compatible = "delta-7448-cpld";
+			reg = <0x1 0x0 0x0000100>;
+		};
+	};
+
+	soc@ffe00000 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		device_type = "soc";
+		compatible = "fsl,p2020-immr", "simple-bus";
+		ranges = <0x0  0x0 0xffe00000 0x100000>;
+		bus-frequency = <0>;
+
+		ecm-law@0 {
+			compatible = "fsl,ecm-law";
+			reg = <0x0 0x1000>;
+			fsl,num-laws = <12>;
+		};
+
+		ecm@1000 {
+			compatible = "fsl,p2020-ecm", "fsl,ecm";
+			reg = <0x1000 0x1000>;
+			interrupts = <17 2>;
+			interrupt-parent = <&mpic>;
+		};
+
+		memory-controller@2000 {
+			compatible = "fsl,p2020-memory-controller";
+			reg = <0x2000 0x1000>;
+			interrupt-parent = <&mpic>;
+			interrupts = <18 2>;
+		};
+
+		i2c@3000 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			cell-index = <0>;
+			compatible = "fsl-i2c";
+			reg = <0x3000 0x100>;
+			interrupts = <43 2>;
+			interrupt-parent = <&mpic>;
+			dfsrr;
+
+			dimm@52 {
+				compatible = "at,spd";
+				reg = <0x52>;
+				read-only;
+			};
+
+			eeprom-mb@54 {
+				compatible = "at,24c08";
+				reg = <0x54>;
+				read-only;
+			};
+
+			rtc@68 {
+				compatible = "stm,m41st85";
+				reg = <0x68>;
+			};
+		};
+
+		i2c@3100 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			cell-index = <1>;
+			compatible = "fsl-i2c";
+			reg = <0x3100 0x100>;
+			interrupts = <43 2>;
+			interrupt-parent = <&mpic>;
+			dfsrr;
+
+			fan@1b {
+				compatible = "maxim,max6650";
+				reg = <0x1b>;
+			};
+
+			fan@1f {
+				compatible = "maxim,max6650";
+				reg = <0x1f>;
+			};
+
+			hotswap@40 {
+				compatible = "lt,ltc4215";
+				reg = <0x40>;
+			};
+
+			hotswap@42 {
+				compatible = "lt,ltc4215";
+				reg = <0x42>;
+			};
+
+			fan@48 {
+				compatible = "maxim,max6650";
+				reg = <0x48>;
+			};
+
+			temp@49 {
+				compatible = "ti,tmp75";
+				reg = <0x49>;
+			};
+
+			temp@4a {
+				compatible = "ti,tmp75";
+				reg = <0x4a>;
+			};
+
+			fan@4b {
+				compatible = "maxim,max6650";
+				reg = <0x4b>;
+			};
+
+			temp@4c {
+				compatible = "ti,tmp75";
+				reg = <0x4c>;
+			};
+
+			temp@4d {
+				compatible = "ti,tmp75";
+				reg = <0x4d>;
+			};
+
+			eeprom-psu-1@50 {
+				read-only;
+				compatible = "at,24c08";
+				reg = <0x50>;
+			};
+
+			eeprom-psu-2@54 {
+				read-only;
+				compatible = "at,24c08";
+				reg = <0x54>;
+			};
+		};
+
+		serial0: serial@4500 {
+			cell-index = <0>;
+			device_type = "serial";
+			compatible = "ns16550";
+			reg = <0x4500 0x100>;
+			clock-frequency = <0>;
+			interrupts = <42 2>;
+			interrupt-parent = <&mpic>;
+		};
+
+		serial1: serial@4600 {
+			cell-index = <1>;
+			device_type = "serial";
+			compatible = "ns16550";
+			reg = <0x4600 0x100>;
+			clock-frequency = <0>;
+			interrupts = <42 2>;
+			interrupt-parent = <&mpic>;
+		};
+
+		dma@c300 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			compatible = "fsl,eloplus-dma";
+			reg = <0xc300 0x4>;
+			ranges = <0x0 0xc100 0x200>;
+			cell-index = <1>;
+			dma-channel@0 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x0 0x80>;
+				cell-index = <0>;
+				interrupt-parent = <&mpic>;
+				interrupts = <76 2>;
+			};
+			dma-channel@80 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x80 0x80>;
+				cell-index = <1>;
+				interrupt-parent = <&mpic>;
+				interrupts = <77 2>;
+			};
+			dma-channel@100 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x100 0x80>;
+				cell-index = <2>;
+				interrupt-parent = <&mpic>;
+				interrupts = <78 2>;
+			};
+			dma-channel@180 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x180 0x80>;
+				cell-index = <3>;
+				interrupt-parent = <&mpic>;
+				interrupts = <79 2>;
+			};
+		};
+
+		gpio: gpio-controller@f000 {
+			#gpio-cells = <2>;
+			compatible = "fsl,mpc8572-gpio";
+			reg = <0xf000 0x100>;
+			interrupts = <47 0x2>;
+			interrupt-parent = <&mpic>;
+			gpio-controller;
+		};
+
+		l2: l2-cache-controller@20000 {
+			compatible = "fsl,p2020-l2-cache-controller";
+			reg = <0x20000 0x1000>;
+			cache-line-size = <32>;
+			cache-size = <0x80000>;
+			interrupt-parent = <&mpic>;
+			interrupts = <16 2>;
+		};
+
+		dma@21300 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			compatible = "fsl,eloplus-dma";
+			reg = <0x21300 0x4>;
+			ranges = <0x0 0x21100 0x200>;
+			cell-index = <0>;
+			dma-channel@0 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x0 0x80>;
+				cell-index = <0>;
+				interrupt-parent = <&mpic>;
+				interrupts = <20 2>;
+			};
+			dma-channel@80 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x80 0x80>;
+				cell-index = <1>;
+				interrupt-parent = <&mpic>;
+				interrupts = <21 2>;
+			};
+			dma-channel@100 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x100 0x80>;
+				cell-index = <2>;
+				interrupt-parent = <&mpic>;
+				interrupts = <22 2>;
+			};
+			dma-channel@180 {
+				compatible = "fsl,eloplus-dma-channel";
+				reg = <0x180 0x80>;
+				cell-index = <3>;
+				interrupt-parent = <&mpic>;
+				interrupts = <23 2>;
+			};
+		};
+
+		ptp_timer: ptimer@24e00 {
+			compatible = "fsl,gianfar-ptp-timer";
+			reg = <0x24e00 0xb0>;
+		};
+
+		enet2: ethernet@26000 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			cell-index = <1>;
+			device_type = "network";
+			model = "eTSEC";
+			compatible = "gianfar";
+			reg = <0x26000 0x1000>;
+			ranges = <0x0 0x26000 0x1000>;
+			local-mac-address = [ 00 00 00 00 00 00 ];
+			interrupts = <31 2 32 2 33 2>;
+			interrupt-parent = <&mpic>;
+			phy-connection-type = "sgmii";
+			phy-handle = <&phy0>;
+			tbi-handle = <&tbi0>;
+		};
+
+		mdio@24520 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			compatible = "fsl,gianfar-mdio";
+			reg = <0x24520 0x20>;
+
+			phy0: ethernet-phy@0 {
+				reg = <0x1>;
+				device_type = "ethernet-phy";
+			};
+			tbi0: tbi-phy@11 {
+				reg = <0x11>;
+				device_type = "tbi-phy";
+			};
+		};
+
+		sdhci@2e000 {
+			compatible = "fsl,p2020-esdhc", "fsl,esdhc";
+			reg = <0x2e000 0x1000>;
+			interrupts = <72 0x2>;
+			interrupt-parent = <&mpic>;
+			clock-frequency = <0>;
+            sdhci,auto-cmd12;
+		};
+
+		crypto@30000 {
+			compatible = "fsl,sec3.1", "fsl,sec3.0", "fsl,sec2.4",
+				     "fsl,sec2.2", "fsl,sec2.1", "fsl,sec2.0";
+			reg = <0x30000 0x10000>;
+			interrupts = <45 2 58 2>;
+			interrupt-parent = <&mpic>;
+			fsl,num-channels = <4>;
+			fsl,channel-fifo-len = <24>;
+			fsl,exec-units-mask = <0xbfe>;
+			fsl,descriptor-types-mask = <0x3ab0ebf>;
+			fsl,multi-host-mode = "dual";
+			fsl,channel-remap = <0x3>;
+		};
+
+		mpic: pic@40000 {
+			interrupt-controller;
+			#address-cells = <0>;
+			#interrupt-cells = <2>;
+			reg = <0x40000 0x40000>;
+			compatible = "chrp,open-pic";
+			device_type = "open-pic";
+		};
+
+		mpic_msgr_block0: message@41400 {
+			compatible = "fsl,mpic-v3.1-msgr";
+			reg = <0x41400 0x200>;
+			interrupts = <
+				0xb0 2
+				0xb1 2
+				0xb2 2
+				0xb3 2>;
+			interrupt-parent = <&mpic>;
+		};
+
+		mpic_msgr_block1: message@42400 {
+			compatible = "fsl,mpic-v3.1-msgr";
+			reg = <0x42400 0x200>;
+			interrupts = <
+				0xb4 2
+				0xb5 2
+				0xb6 2
+				0xb7 2>;
+			interrupt-parent = <&mpic>;
+		};
+
+		msi@41600 {
+			compatible = "fsl,p2020-msi", "fsl,mpic-msi";
+			reg = <0x41600 0x80>;
+			msi-available-ranges = <0 0x100>;
+			interrupts = <
+				0xe0 0
+				0xe1 0
+				0xe2 0
+				0xe3 0
+				0xe4 0
+				0xe5 0
+				0xe6 0
+				0xe7 0>;
+			interrupt-parent = <&mpic>;
+		};
+
+		global-utilities@e0000 {
+			compatible = "fsl,p2020-guts";
+			reg = <0xe0000 0x1000>;
+		};
+	};
+
+	pci2: pcie@ffe0a000 {
+		compatible = "fsl,mpc8548-pcie";
+		device_type = "pci";
+		#interrupt-cells = <1>;
+		#size-cells = <2>;
+		#address-cells = <3>;
+		reg = <0 0xffe0a000 0 0x1000>;
+		bus-range = <0 255>;
+		ranges = <0x2000000 0x0 0xc0000000 0 0xc0000000 0x0 0x20000000
+			  0x1000000 0x0 0x00000000 0 0xffc20000 0x0 0x10000>;
+		clock-frequency = <100000000>;
+		interrupt-parent = <&mpic>;
+		interrupts = <26 2>;
+		interrupt-map-mask = <0xf800 0x0 0x0 0x7>;
+		interrupt-map = <
+			/* IDSEL 0x0 */
+			0000 0x0 0x0 0x1 &mpic 0x0 0x1
+			0000 0x0 0x0 0x2 &mpic 0x0 0x1
+			0000 0x0 0x0 0x3 &mpic 0x0 0x1
+			0000 0x0 0x0 0x4 &mpic 0x0 0x1
+			>;
+		pcie@0 {
+			reg = <0x0 0x0 0x0 0x0 0x0>;
+			#size-cells = <2>;
+			#address-cells = <3>;
+			device_type = "pci";
+			ranges = <0x2000000 0x0 0xc0000000
+				  0x2000000 0x0 0xc0000000
+				  0x0 0x20000000
+
+				  0x1000000 0x0 0x0
+				  0x1000000 0x0 0x0
+				  0x0 0x10000>;
+		};
+	};
+};
diff --git a/arch/powerpc/platforms/85xx/delta_7448.c b/arch/powerpc/platforms/85xx/delta_7448.c
new file mode 100644
index 0000000..4618bef
--- /dev/null
+++ b/arch/powerpc/platforms/85xx/delta_7448.c
@@ -0,0 +1,77 @@
+/*
+ * Delta ET-7448BF platform setup
+ *
+ * Copyright 2013 Big Switch Networks, Inc.
+ *
+ * This program is free software; you can redistribute  it and/or modify it
+ * under  the terms of  the GNU General  Public License as published by the
+ * Free Software Foundation;  either version 2 of the  License, or (at your
+ * option) any later version.
+ */
+
+#include <linux/pci.h>
+#include <asm/udbg.h>
+#include <asm/mpic.h>
+#include <sysdev/fsl_soc.h>
+#include <sysdev/fsl_pci.h>
+
+#include "mpc85xx.h"
+#include "smp.h"
+
+static void delta_7448_restart(char *cmd)
+{
+    struct device_node *cpld;
+
+    cpld = of_find_compatible_node(NULL, NULL, "delta-7448-cpld");
+    if (cpld) {
+        uint8_t __iomem *cpld_regs = of_iomap(cpld, 0);
+        of_node_put(cpld);
+        if (cpld_regs) {
+            writeb(0, cpld_regs + 0x20);
+        }
+    }
+
+    fsl_rstcr_restart(NULL);
+}
+
+static void __init delta_7448_pic_init(void)
+{
+    struct mpic *mpic = mpic_alloc(NULL, 0, MPIC_BIG_ENDIAN, 0, 256,
+                                       " OpenPIC  ");
+    BUG_ON(!mpic);
+    mpic_init(mpic);
+}
+
+static void __init delta_7448_setup_arch(void)
+{
+    if (ppc_md.progress)
+        ppc_md.progress("delta_7448_setup_arch()", 0);
+    fsl_pci_assign_primary();
+    mpc85xx_smp_init();
+}
+
+static void delta_7448_show_cpuinfo(struct seq_file *m)
+{
+    seq_printf(m, "PVR\t\t: 0x%lx\n", mfspr(SPRN_PVR));
+    seq_printf(m, "SVR\t\t: 0x%lx\n", mfspr(SPRN_SVR));
+    seq_printf(m, "PLL\t\t: 0x%lx\n", (mfspr(SPRN_HID1) >> 24) & 0x3f);
+}
+
+static int __init delta_7448_probe(void)
+{
+    return of_flat_dt_is_compatible(of_get_flat_dt_root(), "delta-7448");
+}
+
+machine_arch_initcall(delta_7448, mpc85xx_common_publish_devices);
+
+define_machine(delta_7448) {
+    .name           = "Delta ET-7448BF",
+    .probe          = delta_7448_probe,
+    .setup_arch     = delta_7448_setup_arch,
+    .init_IRQ       = delta_7448_pic_init,
+    .show_cpuinfo   = delta_7448_show_cpuinfo,
+    .get_irq        = mpic_get_irq,
+    .restart        = delta_7448_restart,
+    .calibrate_decr = generic_calibrate_decr,
+    .progress       = udbg_progress,
+};
