--- a/drivers/tty/moxa.c
+++ b/drivers/tty/moxa.c
@@ -867,13 +867,8 @@ static int moxa_init_board(struct moxa_board_conf *brd, struct device *dev)
 	}
 
 	ret = request_firmware(&fw, file, dev);
-	if (ret) {
-		printk(KERN_ERR "MOXA: request_firmware failed. Make sure "
-				"you've placed '%s' file into your firmware "
-				"loader directory (e.g. /lib/firmware)\n",
-				file);
+	if (ret)
 		goto err_free;
-	}
 
 	ret = moxa_load_fw(brd, fw);
 
diff --git a/drivers/tty/serial/icom.c b/drivers/tty/serial/icom.c
index d55709a..0ee5294 100644
