--- a/include/linux/dmi.h
+++ b/include/linux/dmi.h
@@ -93,6 +93,14 @@ struct dmi_dev_onboard {
 	int devfn;
 };
 
+/*
+ * bwh: These functions are renamed so that modules won't load on an
+ * older kernel version where setting the 'exact_match' flag would
+ * result in an array overrun.
+ */
+#define dmi_check_system dmi_check_system_2
+#define dmi_first_match dmi_first_match_2
+
 extern int dmi_check_system(const struct dmi_system_id *list);
 const struct dmi_system_id *dmi_first_match(const struct dmi_system_id *list);
 extern const char * dmi_get_system_info(int field);
diff --git a/include/linux/dynamic_queue_limits.h b/include/linux/dynamic_queue_limits.h
new file mode 100644
index 0000000..5621547
