diff -uNpr a/fs/overlayfs/inode.c b/fs/overlayfs/inode.c
--- a/fs/overlayfs/inode.c	2015-04-07 13:55:17.487864270 -0700
+++ b/fs/overlayfs/inode.c	2015-04-07 13:56:24.948259512 -0700
@@ -68,7 +68,7 @@ int ovl_permission(struct inode *inode,
 			spin_unlock(&inode->i_lock);
 			return -ENOENT;
 		}
-		alias = list_entry(inode->i_dentry.next, struct dentry, d_alias);
+		alias = list_entry(inode->i_dentry.next, struct dentry, d_u.d_alias);
 		dget(alias);
 		spin_unlock(&inode->i_lock);
 		oe = alias->d_fsdata;
