--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -3496,7 +3496,7 @@ void check_move_unevictable_pages(struct page **pages, int nr_pages)
 
 static void warn_scan_unevictable_pages(void)
 {
-	printk_once(KERN_WARNING
+	printk_once(KERN_DEBUG
 		    "The scan_unevictable_pages sysctl/node-interface has been "
 		    "disabled for lack of a legitimate use case.  If you have "
 		    "one, please send an email to linux-mm@kvack.org.\n");
diff --git a/net/8021q/Kconfig b/net/8021q/Kconfig
index fa073a5..4232018 100644
