Fix PPC toology macros

diff --git a/arch/powerpc/include/asm/topology.h b/arch/powerpc/include/asm/topology.h
index c971858..eefc7df 100644
--- a/arch/powerpc/include/asm/topology.h
+++ b/arch/powerpc/include/asm/topology.h
@@ -124,13 +124,15 @@ static inline int stop_topology_update(void)
 #include <asm/cputable.h>
 #define smt_capable()		(cpu_has_feature(CPU_FTR_SMT))
 
-#ifdef CONFIG_PPC64
 #include <asm/smp.h>
 
-#define topology_thread_cpumask(cpu)	(per_cpu(cpu_sibling_map, cpu))
-#define topology_core_cpumask(cpu)	(per_cpu(cpu_core_map, cpu))
-#define topology_core_id(cpu)		(cpu_to_core_id(cpu))
+#ifdef CONFIG_PPC64
+#define topology_thread_cpumask(cpu)		(per_cpu(cpu_sibling_map, cpu))
+#define topology_physical_package_id(cpu)	(get_hard_smp_processor_id(cpu))
 #endif
+#define topology_core_cpumask(cpu)		(per_cpu(cpu_core_map, cpu))
+#define topology_core_id(cpu)			(cpu_to_core_id(cpu))
+#define topology_physical_package_id(cpu)	(0)
 #endif
 
 #endif /* __KERNEL__ */
diff --git a/arch/powerpc/kernel/setup-common.c b/arch/powerpc/kernel/setup-common.c
index 77bb77d..efd43dd 100644
--- a/arch/powerpc/kernel/setup-common.c
+++ b/arch/powerpc/kernel/setup-common.c
@@ -199,6 +199,23 @@ static void show_cpuinfo_summary(struct seq_file *m)
 #endif
 }
 
+/*
+ *	Get CPU information for use by the procfs.
+ */
+static void show_cpuinfo_core(struct seq_file *m, unsigned int cpu)
+{
+
+
+#ifdef CONFIG_SMP
+#ifdef CONFIG_PPC64
+	seq_printf(m, "physical id\t: %d\n", get_hard_smp_processor_id(cpu));
+#else /* Assume 32 bit archs are single package */
+	seq_printf(m, "physical id\t: %d\n", (0));
+#endif /* CONFIG_PPC64 */
+	seq_printf(m, "core id\t\t: %d\n", (cpu_to_core_id(cpu)));
+#endif
+}
+
 static int show_cpuinfo(struct seq_file *m, void *v)
 {
 	unsigned long cpu_id = (unsigned long)v - 1;
@@ -302,6 +319,8 @@ static int show_cpuinfo(struct seq_file *m, void *v)
 	seq_printf(m, "revision\t: %hd.%hd (pvr %04x %04x)\n",
 		   maj, min, PVR_VER(pvr), PVR_REV(pvr));
 
+	show_cpuinfo_core(m, cpu_id);
+
 #ifdef CONFIG_PPC32
 	seq_printf(m, "bogomips\t: %lu.%02lu\n",
 		   loops_per_jiffy / (500000/HZ),
