--- a/arch/x86/kernel/reboot.c	2016-06-01 13:19:53.102025910 -0700
+++ b/arch/x86/kernel/reboot.c	2016-06-01 13:21:04.250027454 -0700
@@ -672,13 +672,13 @@ static void native_machine_emergency_res
 			break;
 
 		case BOOT_CF9_COLD:
-			cf9_cold = inb(0xcf9) & ~6;
-			outb(cf9_cold|8, 0xcf9); /* Request cold reset */
-			udelay(50);
-			outb(cf9_cold|12, 0xcf9); /* Actually do the reset */
-			udelay(50);
-			reboot_type = BOOT_KBD;
-			break;
+                    cf9_cold = inb(0xcf9) & ~0xE;
+                    outb(cf9_cold|2, 0xcf9); /* Request cold reset */
+                    udelay(50);
+                    outb(cf9_cold|0xE, 0xcf9); /* Actually do the reset */
+                    udelay(50);
+                    reboot_type = BOOT_KBD;
+                    break;
 		}
 	}
 }
