Patch to reboot x86_64 machines using boot code CF9

diff --git a/arch/x86/include/asm/emergency-restart.h b/arch/x86/include/asm/emergency-restart.h
index cc70c1c..441a910 100644
--- a/arch/x86/include/asm/emergency-restart.h
+++ b/arch/x86/include/asm/emergency-restart.h
@@ -11,6 +11,7 @@ enum reboot_type {
 	BOOT_EFI = 'e',
 	BOOT_CF9 = 'p',
 	BOOT_CF9_COND = 'q',
+	BOOT_CF9_COLD = 'd',
 };
 
 extern enum reboot_type reboot_type;
diff --git a/arch/x86/kernel/reboot.c b/arch/x86/kernel/reboot.c
index f411aca..a10547c 100644
--- a/arch/x86/kernel/reboot.c
+++ b/arch/x86/kernel/reboot.c
@@ -93,6 +93,7 @@ static int __init reboot_setup(char *str)
 		case 'b':
 #endif
 		case 'a':
+		case 'd':
 		case 'k':
 		case 't':
 		case 'e':
@@ -395,6 +396,16 @@ static int __init set_pci_reboot(const struct dmi_system_id *d)
 	return 0;
 }
 
+static int __init set_cold_cf9_reboot(const struct dmi_system_id *d)
+{
+	if (reboot_type != BOOT_CF9_COLD) {
+		reboot_type = BOOT_CF9_COLD;
+		printk(KERN_INFO "%s series board detected. "
+		       "Selecting CF9 Cold Reset method for reboots.\n", d->ident);
+	}
+	return 0;
+}
+
 static struct dmi_system_id __initdata pci_reboot_dmi_table[] = {
 	{	/* Handle problems with rebooting on Apple MacBook5 */
 		.callback = set_pci_reboot,
@@ -468,6 +479,23 @@ static struct dmi_system_id __initdata pci_reboot_dmi_table[] = {
 			DMI_MATCH(DMI_PRODUCT_NAME, "Precision M6600"),
 		},
 	},
+	{ /* Perform cold reset on Dell S6000 */
+		.callback = set_cold_cf9_reboot,
+		.ident = "Dell S6000",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Dell Inc"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "S6000"),
+		},
+	},
+	{ /* Dell S6000 could have either manufacturer names: Dell Inc or
+	     Dell Force10 Networks. Accomodating for both */
+		.callback = set_cold_cf9_reboot,
+		.ident = "Dell S6000",
+		.matches = {
+			DMI_MATCH(DMI_SYS_VENDOR, "Dell Force10 Networks"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "S6000"),
+		},
+	},
 	{	/* Handle problems with rebooting on the Dell PowerEdge C6100. */
 		.callback = set_pci_reboot,
 		.ident = "Dell PowerEdge C6100",
@@ -569,6 +597,7 @@ static void native_machine_emergency_restart(void)
 	int i;
 	int attempt = 0;
 	int orig_reboot_type = reboot_type;
+	u8 cf9_cold = 0;
 
 	if (reboot_emergency)
 		emergency_vmx_disable_all();
@@ -641,6 +670,15 @@ static void native_machine_emergency_restart(void)
 			}
 			reboot_type = BOOT_KBD;
 			break;
+
+		case BOOT_CF9_COLD:
+			cf9_cold = inb(0xcf9) & ~6;
+			outb(cf9_cold|8, 0xcf9); /* Request cold reset */
+			udelay(50);
+			outb(cf9_cold|12, 0xcf9); /* Actually do the reset */
+			udelay(50);
+			reboot_type = BOOT_KBD;
+			break;
 		}
 	}
 }
