Patch for debian post install script

diff --git a/cumulus/control.linux-headers b/cumulus/control.linux-headers
new file mode 100644
index 0000000..6d999dd
--- /dev/null
+++ b/cumulus/control.linux-headers
@@ -0,0 +1,5 @@
+Package: linux-headers
+Version: =V
+Maintainer: support@cumulusnetworks.com
+Architecture: =A
+Description: Cumulus Linux kernel headers
diff --git a/cumulus/control.linux-image b/cumulus/control.linux-image
new file mode 100644
index 0000000..03fccc4
--- /dev/null
+++ b/cumulus/control.linux-image
@@ -0,0 +1,5 @@
+Package: linux-image
+Version: =V
+Maintainer: support@cumulusnetworks.com
+Architecture: =A
+Description: Cumulus Linux kernel
diff --git a/cumulus/control.linux-libc-dev b/cumulus/control.linux-libc-dev
new file mode 100644
index 0000000..4ae15ce
--- /dev/null
+++ b/cumulus/control.linux-libc-dev
@@ -0,0 +1,12 @@
+Package: linux-libc-dev
+Source: linux-upstream
+Version: =V
+Architecture: =A
+Maintainer: cumulus <support@cumulusnetworks.com>
+Provides: linux-kernel-headers
+Section: devel
+Priority: optional
+Homepage: http://www.kernel.org/
+Description: Linux support headers for userspace development
+ This package provides userspaces headers from the Linux kernel.  These headers
+ are used by the installed headers for GNU glibc and other system libraries.
diff --git a/cumulus/etc/kernel/postinst.d/update-cumulus b/cumulus/etc/kernel/postinst.d/update-cumulus
new file mode 100755
index 0000000..0333980
--- /dev/null
+++ b/cumulus/etc/kernel/postinst.d/update-cumulus
@@ -0,0 +1,52 @@
+#!/bin/sh -e
+
+version="$1"
+
+# exit if we dont need to update the kernel and initrd
+if [ "$UPDATE_CUMULUS" = 'No' ]; then
+        exit 0
+fi
+
+. /usr/share/cumulus/img/functions
+
+# passing the kernel version is required
+if [ -z "${version}" ]; then
+	echo >&2 "W: cumulus-kernel: ${DPKG_MAINTSCRIPT_PACKAGE:-kernel package} did not pass a version number"
+	exit 2
+fi
+
+bootdir="/boot"
+kernel_newimg_src="${bootdir}/vmlinuz-${version}"
+initrd_newimg_src="${bootdir}/initrd.img-${version}"
+
+if [ ! -e "$kernel_newimg_src" ]; then
+	echo >&2 "W: cumulus-kernel: ${DPKG_MAINTSCRIPT_PACKAGE:-kernel package} cannot find kernel image"
+	exit 2
+fi
+
+if [ ! -e "$initrd_newimg_src" ]; then
+	echo >&2 "W: cumulus-kernel: ${DPKG_MAINTSCRIPT_PACKAGE:-kernel package} cannot find initrd image"
+	exit 2
+fi
+
+cl_check_env root-priv
+
+currslot=${slot_prefix}${active}
+kernel_oldimg=${bootdir}/${kernel_prefix}vmlinuz-*-${currslot}
+initrd_oldimg=${bootdir}/${kernel_prefix}initrd.img-*-${currslot}
+
+kernel_newimg_dst=${bootdir}/${kernel_prefix}vmlinuz-${version}-${currslot}
+initrd_newimg_dst=${bootdir}/${kernel_prefix}initrd.img-${version}-${currslot}
+
+#remove existing currslot images
+rm -f ${kernel_oldimg}
+rm -f ${initrd_oldimg}
+
+# Move kernel/initrd files to the kernel initrd dst files
+cmd="mv -f $kernel_newimg_src $kernel_newimg_dst"
+echo "$cmd"
+$cmd
+
+cmd="mv -f $initrd_newimg_src $initrd_newimg_dst"
+echo "$cmd"
+$cmd
diff --git a/cumulus/postinst.linux-image.powerpc b/cumulus/postinst.linux-image.powerpc
new file mode 100755
index 0000000..2aabfea
--- /dev/null
+++ b/cumulus/postinst.linux-image.powerpc
@@ -0,0 +1,57 @@
+#!/bin/sh
+# postinst script for switchd
+#
+# see: dh_installdeb(1)
+
+set -e
+
+# summary of how this script can be called:
+#        * <postinst> `configure' <most-recently-configured-version>
+#        * <old-postinst> `abort-upgrade' <new version>
+#        * <conflictor's-postinst> `abort-remove' `in-favour' <package>
+#          <new-version>
+#        * <postinst> `abort-remove'
+#        * <deconfigured's-postinst> `abort-deconfigure' `in-favour'
+#          <failed-install-package> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see http://www.debian.org/doc/debian-policy/ or
+# the debian-policy package
+
+
+case "$1" in
+    configure)
+	;;
+
+    abort-upgrade|abort-remove|abort-deconfigue)
+	exit 1
+	;;
+    *)
+	echo "postinst called with unknown argument \`$1'" >&2
+	exit 1
+	;;
+esac
+
+. /usr/share/cumulus/img/functions
+
+install_file=/usr/share/cumulus/.installer/kernel_install.sh
+kimage_linkname="uImage.itb"
+kimage_name="uImage-=V-=A.itb"
+kimage_path="/boot/"$kimage_name
+
+if [ -e "/.buildroot" ]; then
+	cd /boot && ln -sf $kimage_name $kimage_linkname
+	exit 0
+fi
+
+cl_check_env root-priv
+slot=$active
+
+sh $install_file $kimage_path $slot || {
+    log_failure_msg "Kernel installtion failed."
+    log_failure_msg "Problems installing image file."
+    exit 1
+}
+
+cd /boot && ln -sf $kimage_name $kimage_linkname
+
+exit 0
