--- a/drivers/i2c/muxes/Kconfig	2015-10-27 16:54:02.741944172 +0000
+++ b/drivers/i2c/muxes/Kconfig	2015-10-27 16:55:02.645945471 +0000
@@ -56,6 +56,14 @@ config I2C_MUX_PCA954x
 	  This driver can also be built as a module.  If so, the module
 	  will be called pca954x.
 
+config I2C_MUX_PCA954X_DESELECT_ON_EXIT
+       bool "Enable deselect-on-exit feature for PCA954X devices."
+       depends on I2C_MUX_PCA954x
+       help
+          If you say yes here you enable the deselect-on-exit feature in
+          the pca954x i2c driver.
+
+
 config I2C_MUX_DNI_6448
 	tristate "Delta Networks 6448 I2C Mux"
 	depends on EXPERIMENTAL
--- linux-3.2.65-1+deb7u2/drivers/i2c/muxes/pca954x.c	2015-10-27 17:13:03.937968935 +0000
+++ b/drivers/i2c/muxes/pca954x.c	2015-10-27 17:12:41.265968443 +0000
@@ -222,6 +222,10 @@ static int pca954x_probe(struct i2c_clie
 		deselect_on_exit = 1;
 	}
 
+#ifdef CONFIG_I2C_MUX_PCA954X_DESELECT_ON_EXIT
+        deselect_on_exit = 1;
+#endif
+
 	data->type = id->driver_data;
 	data->last_chan = 0;		   /* force the first selection */
 
