patch for celestica Redstone XP

diff --git a/drivers/misc/eeprom/at24.c b/drivers/misc/eeprom/at24.c
index 2ec05ce..841631c 100644
--- a/drivers/misc/eeprom/at24.c
+++ b/drivers/misc/eeprom/at24.c
@@ -247,7 +247,14 @@ static ssize_t at24_eeprom_read(struct at24_data *at24, char *buf,
 			}
 			break;
 		case I2C_SMBUS_BYTE_DATA:
-			status = i2c_smbus_read_byte_data(client, offset);
+			if (at24->chip.flags & AT24_FLAG_ADDR16) {
+				status = i2c_smbus_write_byte_data(client, (offset >> 8) & 0xff, offset & 0xff);
+				if (status >= 0) {
+					status = i2c_smbus_read_byte(client);
+				}
+			} else {
+				status = i2c_smbus_read_byte_data(client, offset);
+			}
 			if (status >= 0) {
 				buf[0] = status;
 				status = count;
@@ -405,8 +412,11 @@ static ssize_t at24_eeprom_write(struct at24_data *at24, const char *buf,
 				status = count;
 			break;
 		case I2C_SMBUS_BYTE_DATA:
-			status = i2c_smbus_write_byte_data(client, offset,
-							   buf[0]);
+			if (at24->chip.flags & AT24_FLAG_ADDR16) {
+				status = i2c_smbus_write_word_data(client, (offset >> 8) & 0xff, buf[0] << 8 | (offset & 0xff));
+			} else {
+				status = i2c_smbus_write_byte_data(client, offset,  buf[0]);
+			}
 			if (status == 0)
 				status = count;
 			break;
@@ -571,10 +581,8 @@ static int at24_probe(struct i2c_client *client, const struct i2c_device_id *id)
 	/* Use I2C operations unless we're stuck with SMBus extensions. */
 	if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C)) {
 		if (chip.flags & AT24_FLAG_ADDR16) {
-			err = -EPFNOSUPPORT;
-			goto err_out;
-		}
-		if (i2c_check_functionality(client->adapter,
+			use_smbus = I2C_SMBUS_BYTE_DATA;
+		} else if (i2c_check_functionality(client->adapter,
 				I2C_FUNC_SMBUS_READ_I2C_BLOCK)) {
 			use_smbus = I2C_SMBUS_I2C_BLOCK_DATA;
 		} else if (i2c_check_functionality(client->adapter,
