--- a/drivers/hwmon/adm1021.c	2014-12-14 08:24:02.000000000 -0800
+++ b/drivers/hwmon/adm1021.c	2016-10-13 10:48:10.045055678 -0700
@@ -105,6 +105,7 @@ static struct adm1021_data *adm1021_upda
 /* (amalysh) read only mode, otherwise any limit's writing confuse BIOS */
 static int read_only;
 
+static int detect = 1;
 
 static const struct i2c_device_id adm1021_id[] = {
 	{ "adm1021", adm1021 },
@@ -295,6 +296,9 @@ static int adm1021_detect(struct i2c_cli
 			 "smbus byte data not supported!\n");
 		return -ENODEV;
 	}
+        if(detect == 0) {
+            return -ENODEV;
+        }
 
 	status = i2c_smbus_read_byte_data(client, ADM1021_REG_STATUS);
 	conv_rate = i2c_smbus_read_byte_data(client,
@@ -510,6 +514,8 @@ MODULE_LICENSE("GPL");
 
 module_param(read_only, bool, 0);
 MODULE_PARM_DESC(read_only, "Don't set any values, read only mode");
+module_param(detect, bool, 1);
+MODULE_PARM_DESC(detect, "Enable or disable device detection.");
 
 module_init(sensors_adm1021_init)
 module_exit(sensors_adm1021_exit)
