parse and propagate errors to the user.

diff --git a/net/core/port.c b/net/core/port.c
index e931617..e13c26b 100644
--- a/net/core/port.c
+++ b/net/core/port.c
@@ -359,6 +359,7 @@ struct wq {
 	int hit;
 	int (*decode)(struct nlattr **attrs, void *arg1, void *arg2);
 	void *arg1, *arg2;
+	int status;
 	struct list_head list;
 };
 
@@ -424,10 +425,12 @@ static int port_wait(int seq, int wait,
 	if (err == 0) {	/* timed out */
 		err = -ETIMEDOUT;
 		goto err_out;
+	} else if (wq->status) {
+		err = wq->status;
+	} else {
+		err = 0;
 	}
 
-	err = 0;
-
 err_out:
 	write_lock(&port_wq_lock);
 	list_del(&wq->list);
@@ -492,8 +495,12 @@ static int port_reply(struct sk_buff *skb, struct genl_info *info)
 		wq->hit = 1;
 		if (info->attrs[PORT_ATTR_STATUS]) {
 			err = nla_get_u32(info->attrs[PORT_ATTR_STATUS]);
-			if (!err && wq->decode)
-				wq->decode(info->attrs, wq->arg1, wq->arg2);
+			if (!err) {
+				if (wq->decode)
+					wq->decode(info->attrs, wq->arg1, wq->arg2);
+			} else {
+				wq->status = err;
+			}
 		}
 		wake_up_interruptible(&wq->wq);
 	}
