Remove the carrier_ok check when setting STP state

diff --git a/net/bridge/br_netlink.c b/net/bridge/br_netlink.c
index 79f8e75..9c318a5 100644
--- a/net/bridge/br_netlink.c
+++ b/net/bridge/br_netlink.c
@@ -183,9 +183,8 @@ static int br_rtm_setlink(struct sk_buff *skb,  struct nlmsghdr *nlh, void *arg)
 	if (p->br->stp_enabled == BR_KERNEL_STP)
 		return -EBUSY;
 
-	if (!netif_running(dev) ||
-	    (!netif_carrier_ok(dev) && new_state != BR_STATE_DISABLED))
-		return -ENETDOWN;
+	if (p->state == new_state)
+		return 0;
 
 	p->state = new_state;
 	br_log_state(p);
