vxlan fixes

diff --git a/net/core/rtnetlink.c b/net/core/rtnetlink.c
index 329bb72..c2d619b 100644
--- a/net/core/rtnetlink.c
+++ b/net/core/rtnetlink.c
@@ -2115,7 +2115,10 @@ static int rtnl_fdb_add(struct sk_buff *skb, struct nlmsghdr *nlh, void *arg)
                                                    nlh->nlmsg_flags);
 
                 if (!err) {
-                        rtnl_fdb_notify(dev, addr, dst, RTM_NEWNEIGH);
+                        /* In all cases the netdev fdb_add handler
+			 * sends a netlink notification. So we dont need
+			 * it here again.
+			 * rtnl_fdb_notify(dev, addr, dst, RTM_NEWNEIGH); */
                         ndm->ndm_flags &= ~NTF_SELF;
                 }
         }
@@ -2175,7 +2178,10 @@ static int rtnl_fdb_del(struct sk_buff *skb, struct nlmsghdr *nlh, void *arg)
                 err = dev->netdev_ops->ndo_fdb_del(ndm, dev, addr);
 
                 if (!err) {
-                        rtnl_fdb_notify(dev, addr, NULL, RTM_DELNEIGH);
+                        /* In all cases the netdev fdb_add handler
+			 * sends a netlink notification. So we dont need
+			 * it here again.
+			 * rtnl_fdb_notify(dev, addr, NULL, RTM_DELNEIGH); */
                         ndm->ndm_flags &= ~NTF_SELF;
                 }
         }
