From 5d87993c403c54852c9e7a6ed71eba1295ebf903 Mon Sep 17 00:00:00 2001
Subject: [PATCH] cumulus: bonding: netlink dump lacp fallback attributes

Adds support for dumping the various lacp fallback attributes.

diff --git a/drivers/net/bonding/bond_main.c b/drivers/net/bonding/bond_main.c
index 14b8bc1..3891c08 100644
--- a/drivers/net/bonding/bond_main.c
+++ b/drivers/net/bonding/bond_main.c
@@ -5068,6 +5068,9 @@ static size_t bond_get_size(const struct net_device *bond_dev)
 		nla_total_size(sizeof(u16)) + /* IFLA_BOND_AD_INFO_ACTOR_KEY */
 		nla_total_size(sizeof(u16)) + /* IFLA_BOND_AD_INFO_PARTNER_KEY*/
 		nla_total_size(ETH_ALEN) +    /* IFLA_BOND_AD_INFO_PARTNER_MAC*/
+		nla_total_size(sizeof(u8)) +  /* IFLA_BOND_CL_LACP_FB_ALLOW */
+		nla_total_size(sizeof(u8)) +  /* IFLA_BOND_CL_LACP_FB_ACTIVE */
+		nla_total_size(sizeof(u32)) + /* IFLA_BOND_CL_LACP_FB_PERIOD */
 		0;
 }
 
@@ -5214,6 +5217,18 @@ static int bond_fill_info(struct sk_buff *skb,
 
 			nla_nest_end(skb, nest);
 		}
+
+		if (nla_put_u8(skb, IFLA_BOND_CL_LACP_FB_ALLOW,
+			       bond->params.lacp_fallback_allow))
+		    goto nla_put_failure;
+
+		if (nla_put_u8(skb, IFLA_BOND_CL_LACP_FB_ACTIVE,
+			       bond->params.lacp_fallback_active))
+		    goto nla_put_failure;
+
+		if (nla_put_u32(skb, IFLA_BOND_CL_LACP_FB_PERIOD,
+			        bond->params.lacp_fallback_period))
+		    goto nla_put_failure;
 	}
 
 	return 0;
@@ -5231,6 +5246,7 @@ static size_t bond_get_slave_size(const struct net_device *bond_dev,
 		nla_total_size(MAX_ADDR_LEN) +	/* IFLA_BOND_SLAVE_PERM_HWADDR */
 		nla_total_size(sizeof(u16)) +	/* IFLA_BOND_SLAVE_QUEUE_ID */
 		nla_total_size(sizeof(u16)) +	/* IFLA_BOND_SLAVE_AD_AGGREGATOR_ID */
+		nla_total_size(sizeof(u32)) +   /* IFLA_BOND_SLAVE_CL_LACP_FB_PRIO */
 		0;
 }
 
@@ -5268,6 +5284,10 @@ static int bond_fill_slave_info(struct sk_buff *skb,
 			if (nla_put_u16(skb, IFLA_BOND_SLAVE_AD_AGGREGATOR_ID,
 					agg->aggregator_identifier))
 				goto nla_put_failure;
+
+		if (nla_put_u32(skb, IFLA_BOND_SLAVE_CL_LACP_FB_PRIO,
+				slave->lacp_fallback_priority))
+			goto nla_put_failure;
 	}
 
 	return 0;
diff --git a/include/linux/if_link.h b/include/linux/if_link.h
index 158b989..ee99f97 100644
--- a/include/linux/if_link.h
+++ b/include/linux/if_link.h
@@ -352,6 +352,11 @@ enum {
 	IFLA_BOND_AD_LACP_RATE,
 	IFLA_BOND_AD_SELECT,
 	IFLA_BOND_AD_INFO,
+
+	IFLA_BOND_CL_START = 100,
+	IFLA_BOND_CL_LACP_FB_ALLOW = IFLA_BOND_CL_START,
+	IFLA_BOND_CL_LACP_FB_ACTIVE,
+	IFLA_BOND_CL_LACP_FB_PERIOD,
 	__IFLA_BOND_MAX,
 };
 
@@ -377,6 +382,9 @@ enum {
 	IFLA_BOND_SLAVE_PERM_HWADDR,
 	IFLA_BOND_SLAVE_QUEUE_ID,
 	IFLA_BOND_SLAVE_AD_AGGREGATOR_ID,
+
+	IFLA_BOND_SLAVE_CL_START = 50,
+	IFLA_BOND_SLAVE_CL_LACP_FB_PRIO = IFLA_BOND_SLAVE_CL_START,
 	__IFLA_BOND_SLAVE_MAX,
 };
 
