Add replace command for mdb entries

diff --git a/net/bridge/br_mdb.c b/net/bridge/br_mdb.c
index d186c09..257e0f0 100644
--- a/net/bridge/br_mdb.c
+++ b/net/bridge/br_mdb.c
@@ -531,6 +531,8 @@ static int br_mdb_add(struct sk_buff *skb, struct nlmsghdr *nlh, void *arg)
 		if (!err)
 			__br_mdb_notify(dev, entry, RTM_NEWMDB);
 	}
+	if ((nlh->nlmsg_flags & NLM_F_REPLACE) && (err == -EEXIST))
+		err = 0;
 	return err;
 }
 
