Set the MAX_MTU to 9216

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 0c31f58..8ac0b56 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -459,12 +459,12 @@ static void tun_net_mclist(struct net_device *dev)
 }
 
 #define MIN_MTU 68
-#define MAX_MTU 65535
+#define MAX_MTU 9216
 
 static int
 tun_net_change_mtu(struct net_device *dev, int new_mtu)
 {
-	if (new_mtu < MIN_MTU || new_mtu + dev->hard_header_len > MAX_MTU)
+	if (new_mtu < MIN_MTU || new_mtu > MAX_MTU)
 		return -EINVAL;
 	dev->mtu = new_mtu;
 	return 0;
