debug

diff --git a/net/core/port.c b/net/core/port.c
index ac4fc82..e931617 100644
--- a/net/core/port.c
+++ b/net/core/port.c
@@ -735,7 +735,17 @@ err_out:
 
 int port_get_settings(struct net_device *dev, struct ethtool_cmd *cmd)
 {
-	return port_cache_get_settings(dev->ifindex, cmd);
+	int err;
+
+	err = port_cache_get_settings(dev->ifindex, cmd);
+	/* If we dont have ethtool data from the driver, its better to return zero
+	 * to the caller instead of returning an err. Because this is really
+	 * not an error and sets can really go through at this point.
+	 */
+	if (err == -ENODATA)
+		return 0;
+
+	return err;
 }
 EXPORT_SYMBOL_GPL(port_get_settings);
 
