Tune out the sysctl entries for large vlan count

diff --git a/net/8021q/vlan.c b/net/8021q/vlan.c
index 963f285..66cb2ef 100644
--- a/net/8021q/vlan.c
+++ b/net/8021q/vlan.c
@@ -188,6 +188,8 @@ int register_vlan_dev(struct net_device *dev)
 	if (err < 0)
 		goto out_uninit_applicant;
 
+	dev->vlan_features |= NETIF_F_HIGHDMA;
+
 	err = register_netdevice(dev);
 	if (err < 0)
 		goto out_uninit_applicant;
diff --git a/net/ipv4/devinet.c b/net/ipv4/devinet.c
index f9e6a35..dc49a02 100644
--- a/net/ipv4/devinet.c
+++ b/net/ipv4/devinet.c
@@ -1720,9 +1720,12 @@ static void __devinet_sysctl_unregister(struct ipv4_devconf *cnf)
 
 static void devinet_sysctl_register(struct in_device *idev)
 {
-	neigh_sysctl_register(idev->dev, idev->arp_parms, "ipv4", NULL);
-	__devinet_sysctl_register(dev_net(idev->dev), idev->dev->name,
-					&idev->cnf);
+
+	if (!(idev->dev->vlan_features & NETIF_F_HIGHDMA)) {
+		neigh_sysctl_register(idev->dev, idev->arp_parms, "ipv4", NULL);
+		__devinet_sysctl_register(dev_net(idev->dev), idev->dev->name,
+					  &idev->cnf);
+	}
 }
 
 static void devinet_sysctl_unregister(struct in_device *idev)
