

diff --git a/net/bridge/br_input.c b/net/bridge/br_input.c
index 61f60bb..1c52833 100644
--- a/net/bridge/br_input.c
+++ b/net/bridge/br_input.c
@@ -22,6 +22,7 @@
 /* Bridge group multicast address 802.1d (pg 51). */
 const u8 br_group_address[ETH_ALEN] = { 0x01, 0x80, 0xc2, 0x00, 0x00, 0x00 };
 const u8 br_pvst_address[ETH_ALEN] = { 0x01, 0x00, 0x0c, 0xcc, 0xcc, 0xcd };
+const u8 br_cdp_address[ETH_ALEN] = { 0x01, 0x00, 0x0c, 0xcc, 0xcc, 0xcc };
 
 /* Hook for brouter */
 br_should_route_hook_t __rcu *br_should_route_hook __read_mostly;
@@ -182,6 +183,17 @@ rx_handler_result_t br_handle_frame(struct sk_buff **pskb)
 		}
 	}
 
+	if (!compare_ether_addr(br_cdp_address, dest)) {
+		/* Deliver packet to local host only */
+		if (NF_HOOK(NFPROTO_BRIDGE, NF_BR_LOCAL_IN, skb, skb->dev,
+					NULL, br_handle_local_finish)) {
+			return RX_HANDLER_CONSUMED; /* consumed by filter */
+		} else {
+			*pskb = skb;
+			return RX_HANDLER_PASS;	/* continue processing */
+		}
+	}
+
 	if (unlikely(is_link_local(dest))) {
 		/*
 		 * See IEEE 802.1D Table 7-10 Reserved addresses
