route: Send flags in new route notify

diff --git a/net/ipv4/fib_trie.c b/net/ipv4/fib_trie.c
index 06172d6..10bbea4 100644
--- a/net/ipv4/fib_trie.c
+++ b/net/ipv4/fib_trie.c
@@ -1339,7 +1339,7 @@ int fib_table_insert(struct fib_table *tb, struct fib_config *cfg)
 
 	rt_cache_flush(cfg->fc_nlinfo.nl_net, -1);
 	rtmsg_fib(RTM_NEWROUTE, htonl(key), new_fa, plen, tb->tb_id,
-		  &cfg->fc_nlinfo, 0);
+		  &cfg->fc_nlinfo, cfg->fc_nlflags);
 succeeded:
 	return 0;
 
